---
title: "Differential Expression"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
params:
    bcb_file: "data/2018-02-20/bcb.rda"
    design: !r formula(~date + category)
    contrast: !r c("category", "ALS", "control")
    alpha: 0.05
    lfc: 0
    output_dir: "."
    dropbox_dir: !r basename(getwd())
---

```{r setup, message=FALSE}
# Last modified 2018-02-20
bcbioRNASeq::prepareRNASeqTemplate()
source("_setup.R")

# Directory paths
data_dir <- file.path(params$output_dir, "data", Sys.Date())
results_dir <- file.path(params$output_dir, "results", "differential_expression")
invisible(mapply(
    FUN = dir.create,
    list(data_dir, results_dir),
    MoreArgs = list(recursive = TRUE, showWarnings = FALSE)
))

# Load bcbioRNASeq object
bcb_name <- load(params$bcb_file)
bcb <- get(bcb_name, inherits = FALSE)
stopifnot(is(bcb, "bcbioRNASeq"))
invisible(validObject(bcb))
```

```{r header, child="_header.Rmd"}
```



For the differential expression analysis, we used [DESeq2][] `r packageVersion("DESeq2")`. We employed a design formula (`r paste(params$design, collapse = "")`) that accounts for the date batch effect. For more information on design formulas in [DESeq2][], run `help("design", "DESeq2")` in [R][].

Here we're discarding the NoTx samples from the analysis. Note that since these 3 samples have been dropped, the values of the normalized counts will change, since we're now only looking at 15 samples. I've uploaded these updated counts into the `results/` directory.

```{r dds, results="hide"}
dds <- bcbio(bcb, "DESeqDataSet")
dim(dds)

# Drop the `NoTx_*` samples from the analysis
keep_samples <- colnames(dds)[which(!is.na(dds$treatment))]
keep_samples

dds <- selectSamples(dds, sampleID = keep_samples)
dim(dds)
colnames(dds)

# Now set the design formula and run DESeq
design(dds) <- params$design
dds <- DESeq(dds)
rld <- rlog(dds)
```

Since the dimensions of the dataset have changed, we need to resave the normalized counts.

```{r resave_normalized_counts}
normalized_counts <- counts(dds, normalized = TRUE)
writeCounts(normalized_counts, dir = file.path("results", "counts"))
saveData(dds, normalized_counts, rld, dir = data_dir)
```

```{r dropbox_resave, message=TRUE}
dropbox_normalized_counts_resave <- copyToDropbox(
    c(normalized_counts = file.path("results", "counts", "normalized_counts.csv.gz")),
    dir = params$dropbox_dir)
saveData(dropbox_normalized_counts_resave, dir = data_dir)
```



# Alpha level (FDR) cutoffs

Let's take a look at the number of genes we get with different false discovery rate (FDR) cutoffs. These tests subset *P* values that have been multiple test corrected using the Benjamini Hochberg (BH) method [@Benjamini:1995ws].

```{r alpha_summary, results="asis"}
alphaSummary(dds)
```



# Results

```{r res}
# help("results", "DESeq2")
# For contrast argument as character vector:
#   1. Design matrix factor of interest.
#   2. Numerator for LFC (expt).
#   3. Denominator for LFC (control).
res_unshrunken <- results(
    dds,
    contrast = params$contrast,
    alpha = params$alpha)

# DESeqResults with shrunken log2 fold changes (LFC)
# help("lfcShrink", "DESeq2")
res_shrunken <- lfcShrink(
    dds = dds,
    contrast = params$contrast,
    res = res_unshrunken)

# Use shrunken LFC values by default
res <- res_shrunken
saveData(res, res_shrunken, res_unshrunken, dir = data_dir)
```

We performed the analysis using a BH adjusted *P* value cutoff of `r params$alpha` and a log fold-change (LFC) ratio cutoff of `r params$lfc`.



# Plots

We're seeing a small but consistent effect size. Based on these results, I don't recommend applying a log fold change cutoff to our list of differentially expressed genes.


## Mean average (MA) {.tabset}

An MA plot compares transformed counts on `M` (log ratio) and `A` (mean average) scales [@Yang:2002ty].

```{r plot_ma}
markdownHeader("shrunken", level = 3)
plotMA(res_shrunken)

markdownHeader("unshrunken", level = 3)
plotMA(res_unshrunken)
```


## Volcano {.tabset}

A volcano plot compares significance (BH-adjusted *P* value) against fold change (log2) [@Cui:2003kh; @Li:2014fv]. Genes in the green box with text labels have an adjusted *P* value are likely to be the top candidate genes of interest.

```{r plot_volcano}
markdownHeader("labeled", level = 3)
plotVolcano(
    res,
    lfc = params$lfc,
    gene2symbol = gene2symbol(bcb),
    ntop = 20,
    histograms = TRUE)
```

```{r plot_volcano_additional, include=FALSE}
plotVolcano(res_shrunken, padj = TRUE)
plotVolcano(res_shrunken, padj = FALSE)

plotVolcano(res_unshrunken, padj = TRUE)
plotVolcano(res_unshrunken, padj = FALSE)
```


## Heatmap

This plot shows only differentially expressed genes on a per-sample basis. We have scaled the data by row and used the `ward.D2` method for clustering [@WardJr:1963eu].

```{r plot_deg_heatmap}
# help("pheatmap", "pheatmap")
annotation_col <- colData(dds) %>%
    .[, c("category", "treatment", "date"), drop = FALSE] %>%
    as.data.frame()
plotDEGHeatmap(
    res,
    counts = rld,
    gene2symbol = gene2symbol(bcb),
    annotationCol = annotation_col,
    clustering_method = "ward.D2")
```


# DEG PCA plots {.tabset}

Now performing PCA using only our list of DEGs, we can see the samples are separating by treatment on PC1, which is good to see. The samples are separating on PC2 by date (batch), but that's okay because we've taken that into account in our DESeq model for performing the differential expression. This looks good.

```{r plot_deg_pca}
markdownHeader("category", level = 2)
plotDEGPCA(res, counts = rld, interestingGroups = "category")

markdownHeader("date", level = 2)
plotDEGPCA(res, counts = rld, interestingGroups = "date")

markdownHeader("category:date", level = 2)
plotDEGPCA(res, counts = rld, interestingGroups = c("category", "date"), label = TRUE)
```



# Results tables

```{r results_tables, results="asis"}
resTbl <- resultsTables(
    res,
    lfc = params$lfc,
    annotable = annotable(bcb),
    summary = TRUE,
    headerLevel = 2,
    write = TRUE,
    dir = results_dir,
    dropboxDir = params$dropbox_dir)
saveData(resTbl, dir = data_dir)
```

DEG tables are sorted by BH-adjusted P value, and contain the following columns:

- `ensgene`: [Ensembl][] gene identifier.
- `baseMean`: Mean of the normalized counts per gene for all samples.
- `log2FoldChange`: log2 fold change.
- `lfcSE`: log2 standard error.
- `stat`: Wald statistic.
- `pvalue`: Walt test *P* value.
- `padj`: BH adjusted Wald test *P* value (corrected for multiple comparisons; aka FDR).
- `externalGeneName`: [Ensembl][] name (a.k.a. symbol).
- `description`: [Ensembl][] description.
- `geneBiotype`: [Ensembl][] biotype (e.g. `protein_coding`).


## Top tables

Only the top up- and down-regulated genes (arranged by log2 fold change) are shown.

```{r top_tables, results="asis"}
topTables(resTbl)
```



```{r footer, child="_footer.Rmd"}
```
